### PR TITLE
Construct bases-mask from samplesheet

### PR DESCRIPTION
**Pull request by johandahlberg**
The bases-mask will per default be defined based on the index defined in the samplesheet, on a per lane basis.

Currently this only supports the old style of illumina samplesheets, but we need to make sure it supports the new style as well.

### CODE REVIEW

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 1]
```diff
@@ -179,7 +253,13 @@ def construct_command(self):
...
 
         if self.config.use_base_mask:
-            commandline_collection.append("--use_base_mask " + self.config.use_base_mask)
+            # Note that for the base mask the "--use_base_mask" must be included in the
+            # commandline passed.
+            commandline_collection.append(self.config.use_base_mask)
+        else:
+            lanes_and_base_mask = Bcl2FastqConfig.get_bases_mask_per_lane_from_samplesheet(self.config.samplesheet)
+            for lane, base_mask in lanes_and_base_mask.iteritems():
+                commandline_collection.append("--use_base_mask {0}:{1}".format(lane, base_mask))
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: the bcl2fastq command line option is actually `--use-bases-mask`

Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: Good catch.

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 2]
```diff
@@ -66,8 +70,78 @@ def get_bcl2fastq_version_from_run_parameters(runfolder, config):
...
+        normally parsed from run meta data.
+        :return a dict of the lane and base mask to use, e.g.:
+                 { 1:"y*,iiiiiiiin*,iiiiiiiin*,y*" , 2:"y*,iiiiii,n*,y*  [etc] }
+        """
+        def is_double_index(idx):
+            return "-" in idx
+
+        def pad_with_ignore(length_of_index_in_samplesheet, length_of_index_read):
+            difference = length_of_index_read - length_of_index_in_samplesheet
+            print("difference: " + str(difference))
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: is this a leftover debug statement?

Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: Yes. Missed that.

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 3]
```diff
@@ -179,7 +253,13 @@ def construct_command(self):
             commandline_collection.append("--tiles " + self.config.tiles)
 
         if self.config.use_base_mask:
-            commandline_collection.append("--use_base_mask " + self.config.use_base_mask)
+            # Note that for the base mask the "--use_base_mask" must be included in the
+            # commandline passed.
+            commandline_collection.append(self.config.use_base_mask)
+        else:
+            lanes_and_base_mask = Bcl2FastqConfig.get_bases_mask_per_lane_from_samplesheet(self.config.samplesheet)
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: missing the `index_lengths` argument in the call to `Bcl2FastqConfig.get_bases_mask_per_lane_from_samplesheet`

Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: Also a good catch. I keep expecting a compilation error message for this type of thing...

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 4]
```diff
@@ -66,8 +70,78 @@ def get_bcl2fastq_version_from_run_parameters(runfolder, config):
 
         return version
 
+    @staticmethod
+    def get_length_of_indexes(runfolder):
+        """
+        Will parse runfolder meta data to find the length of the index reads.
+        :param runfolder: to get the length of the index reads from.
+        :return: a cict with the index number as key and the length of each index as value e.g.:
+                 {1: 7, 2: 8}
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: is this actually what this method will return? at least in the `RunInfo.xml`, the reads are incremented from 1, regardless if it's an index read or not. So in e.g. a paired-end dual-indexed run, the index reads will have number 2 and 3.

Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: Aha. I didn't know. I'll fix that up real quick.

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 5]
```diff
@@ -66,8 +70,78 @@ def get_bcl2fastq_version_from_run_parameters(runfolder, config):
...
+        """
+        meta_data = InteropMetadata(runfolder)
+        index_read_info = filter(lambda x: x["is_index"], meta_data.read_config)
+        indexes_and_lengths = map(lambda x: (x["read_num"], x["cycles"]), index_read_info)
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: something like: `indexes_and_lengths = {i:x["cycles"] for i,x in enumerate(sorted(index_read_info,key=lambda x: x["read_num"]))}`

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 6]
```diff
@@ -218,6 +298,15 @@ def construct_command(self):
...
             commandline_collection.append("--use_bases_mask " + self.config.use_base_mask)
+        else:
+            lanes_and_base_mask = Bcl2FastqConfig.get_bases_mask_per_lane_from_samplesheet(self.config.samplesheet)
+            base_masks_as_set = set(lanes_and_base_mask.values())
+
+            assert len(base_masks_as_set) is 1, "For bcl2fastq 1.8.4 there is no support for " \
+                                                "mixing different bases masks for different lanes"
+
+            # Here we are forced to use the same bases mask was always used for all lanes.
+            commandline_collection.append("--use_bases_mask " + lanes_and_base_mask.values()[0])
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: nb: `--use-bases-mask`

File: bcl2fastq/tests/tests_bcl2fastq.py [hunk 7]
```diff
@@ -50,14 +83,15 @@ def test_construct_command(self):
...
+            use_base_mask="--use_base_mask y*,6i,6i,y* --use_base_mask 1:y*,5i,5i,y*",
             additional_args="--my-best-arg 1 --my-best-arg 2")
 
         runner = BCL2Fastq2xRunner(config, "/bcl/binary/path")
         command = runner.construct_command()
         expected_command = "/bcl/binary/path --input-dir test/runfolder/Data/Intensities/BaseCalls " \
                            "--output-dir test/output --barcode-mismatches 2 " \
-                           "--tiles s1,s2,s3 --use_base_mask Y*NN " \
+                           "--tiles s1,s2,s3 " \
+                           "--use_base_mask y*,6i,6i,y* --use_base_mask 1:y*,5i,5i,y* " \
```
Comment (b97pla) [inline bcl2fastq/tests/tests_bcl2fastq.py]: `--use-bases-mask`

File: bcl2fastq/tests/test_bcl2fastq_handlers.py [hunk 8]
```diff
@@ -10,30 +12,21 @@
...
-                                    "binary": "/path/to/bcl2fastq"}}},
-                     "machine_type":
-                         {"MiSeq": {"bcl2fastq_version": "1.8.4"},
-                          "HiSeq X": {"bcl2fastq_version": "2.15.2"},
-                          "HiSeq 2500": {"bcl2fastq_version": "1.8.4"},
-                          "HiSeq 4000": {"bcl2fastq_version": "1.8.4"},
-                          "HiSeq 2000": {"bcl2fastq_version": "1.8.4"},
-                          "NextSeq 500": {"bcl2fastq_version": "1.8.4"}},
-                     "bcl2fastq_logs_path": "/tmp/"}
+    MOCK_BCL2FASTQ_DICT =  {1: "y*,iiiiiiii,iiiiiiii,y*",
```
Comment (b97pla) [inline bcl2fastq/tests/test_bcl2fastq_handlers.py]: should be `y*,i8n*...` etc.

File: bcl2fastq/tests/test_bcl2fastq_handlers.py [hunk 9]
```diff
@@ -45,9 +38,10 @@ def test_start_missing_runfolder_in_body(self):
     def test_start(self):
         # Use mock to ensure that this will run without
         # creating the runfolder.
-        with mock.patch.object(Config, 'load_config', return_value=TestBcl2FastqHandlers.DUMMY_CONFIG),\
-                mock.patch.object(os.path, 'isdir', return_value=True),\
-                mock.patch.object(Bcl2FastqConfig, 'get_bcl2fastq_version_from_run_parameters', return_value="2.15.2"):
+        with mock.patch.object(Config, 'load_config', return_value=TestUtils.DUMMY_CONFIG), \
+             mock.patch.object(os.path, 'isdir', return_value=True), \
+             mock.patch.object(Bcl2FastqConfig, 'get_bcl2fastq_version_from_run_parameters', return_value="2.15.2"), \
+             mock.patch.object(Bcl2FastqConfig, 'get_bases_mask_per_lane_from_samplesheet', return_value=self.MOCK_BCL2FASTQ_DICT):
```
Comment (b97pla) [inline bcl2fastq/tests/test_bcl2fastq_handlers.py]: I've never worked with testing and mocking web app stuff, I'd be very interested if you have time to show me how your tests work

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py [hunk 10]
```diff
@@ -218,6 +298,15 @@ def construct_command(self):
 
         if self.config.use_base_mask:
             commandline_collection.append("--use_bases_mask " + self.config.use_base_mask)
+        else:
+            lanes_and_base_mask = Bcl2FastqConfig.get_bases_mask_per_lane_from_samplesheet(self.config.samplesheet)
```
Comment (b97pla) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: missing argument here as well

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py (hunk 6 above)
Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: :+1:

File: bcl2fastq/tests/tests_bcl2fastq.py (hunk 7 above)
Comment (johandahlberg) [inline bcl2fastq/tests/tests_bcl2fastq.py]: :+1:

File: bcl2fastq/tests/test_bcl2fastq_handlers.py (hunk 8 above)
Comment (johandahlberg) [inline bcl2fastq/tests/test_bcl2fastq_handlers.py]: I this context it doesn't really matter that much. But I've fixed it.

File: bcl2fastq/tests/test_bcl2fastq_handlers.py (hunk 9 above)
Comment (johandahlberg) [inline bcl2fastq/tests/test_bcl2fastq_handlers.py]: I'd be happy to show you.

File: bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py (hunk 10 above)
Comment (johandahlberg) [inline bcl2fastq/bcl2fastq/lib/bcl2fastq_utils.py]: Fixed.

File: bcl2fastq/tests/test_bcl2fastq_handlers.py (hunk 8 above)
Comment (b97pla) [inline bcl2fastq/tests/test_bcl2fastq_handlers.py]: ...and now I see why :smile: it was just my ignorance regarding the web app tests talking

**Pull request merged**